### PR TITLE
NIFI-9069: Changed framework so that it serializes the dataflow into …

### PR DESCRIPTION
**Pull request by markap14**
…a VersionedDataflow using JSON as well as XML, and prefers the JSON representation on load, if it's available. This also results in the need for the cluster protocol to exchange its representation of the dataflow to using JSON. Rather than re-implementing all of the complex logic of Flow Fingerprinting, updated to just inherit the cluster's flow.

- Note that while the following list of changes may at first seem to be a long list of unrelated improvements, refactorings all gathered into a single PR, it needed to be done this way. Changing flow persistence from XML to JSON meant that the Cluster Coordinator will now provide the flow as a JSON instead of an XML. This breaks our Flow Fingerprint mechanism for determining if a node should be allowed to join a cluster. Rather than re-implement all of the logic for Flow Fingerprinting based on JSON it made more sense and is simpler to just use the synchronization logic we already have for dealing with Versioned Components. These changes made it easier then to inherit a cluster flow, even if there are conflicts because this synchronization logic is more robust and includes starting/stopping components as necessary. Because of this change, now nodes will be able to rejoin even with conflicts so it didn't make sense to continue the requirement that the cluster cannot be updated if a node is not part of the cluster, so that check was removed. I looked at separating these changes out into separate Jiras/PRs but doing so would have greatly complicated things because of all of the interdependencies of these tasks, so I kept it all as a single PR. Below is a more comprehensive list of changes/improvements/refactorings that were necessary as part of this changeset:

- Moved logic to synchronize Process Group with Versioned Process Group into a new ProcessGroupSynchronizer class instead of having all of the logic within StandardProcessGroup. This significantly improves readability and avoids importing things into the StandardProcessGroup that are not actually part of the ProcessGroup.
- Reworked versioned components to use an instance id. This is important when synching an existing component to a cluster's component rather than a Versioned Flow.
- Renamed StandardFlowSynchronizer to XmlFlowSynchronizer; introduced new StandardFlowSynchronizer that delegates to the appropriate (Xml or Versioned)FlowSynchronzer
- Updated to allow import of VersionedProcessGroup even if not all bundles are available - will now use ghost components
- Introduced a VersionedDataflow object to hold controller-level services, reporting tasks, parameter contexts, templates, etc.
- Allow mutable requests to be made while nodes are disconnected. Also fixed issue in AbstractPolicyBasedAuthorizer that caused ClassNotFoundException / NoClassDefFoundError if the authorizations were changed and then a node attempts to rejoin the cluster. The Authorizer was attempting to use XmlUtils, which is in nifi-security-utils and so so by madking nifi-security-utils a provided dependency of nifi-framework-api, but this doesn't work, because nifi-framework-api is loaded by a higher-level classloader, so the classloader that loads AbstractPolicyBasedAuthorizer will never have the appropriate classloader to provide nifi-security-utils. Addressed this by copying the code for creating a safe document builder from XmlUtils to AbstractPolicyBasedAuthorizer.

Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Review by exceptionfactory — COMMENTED** [review]
This looks like a great improvement @markap14!

This will take some time to review, but one initial question, what was the reason for removing the `suppressions.xml` from  `nifi-dependency-check-maven`? This optional OWASP dependency check profile uses this file to avoid false positives during dependency vulnerability scanning, so it should be restored.

**Comment by markap14** [comment]
@exceptionfactory d'oh! Thanks. That was unintentional. I saw that during a rebase and thought it was a file that I had created on a different branch and it just wasn't cleaned up (because i hadn't run `git clean`). Just pushed a new commit that rebases against `main` and restores that file. Great catch!

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/flow/PopularVoteFlowElection.java [hunk 1]
```diff
@@ -141,7 +142,7 @@ public synchronized DataFlow castVote(final DataFlow candidate, final NodeIdenti
     }
 
     private String fingerprint(final DataFlow dataFlow) {
-        final String flowFingerprint = fingerprintFactory.createFingerprint(dataFlow.getFlow());
+        final String flowFingerprint = DigestUtils.sha256Hex(dataFlow.getFlow());
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/flow/PopularVoteFlowElection.java:145]: With the change to flow fingerprinting, the `FingerprintFactory` no longer appears to be used in this class or elsewhere. Should the FingerprintFactory class be removed?  At minimum, it looks like the instance can be removed from this class.

**Review by exceptionfactory — COMMENTED** [review]
@markap14 In reviewing the changes to fingerprinting, it appears the `FingerprintFactory` class is no longer used in `PopularVoteFlowElection`.  Given the change in the reconciliation of flow differences, can the `FingerprintFactory` class be removed?

**Comment by exceptionfactory** [comment]
> @markap14 In reviewing the changes to fingerprinting, it appears the `FingerprintFactory` class is no longer used in `PopularVoteFlowElection`. Given the change in the reconciliation of flow differences, can the `FingerprintFactory` class be removed?

In addition to the `FingerprintFactory`, the `SensitiveValueEncoder` and associated implementation could also be removed if the `FingerprintFactory` is no longer necessary.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/groups/StandardProcessGroupSynchronizer.java [hunk 2]
```diff
@@ -0,0 +1,2050 @@
...
+                    .anyMatch(b -> serviceToAddCoordinate.equals(b.getBundleDetails().getCoordinate()));
+
+                if (!bundleExists && possibleBundles.size() != 1) {
+                    LOG.warn("Unknown bundle {} for processor type {} - will use Ghosted component instead", serviceToAddCoordinate, serviceToAddClass);
+//                    throw new IllegalArgumentException("Unknown bundle " + serviceToAddCoordinate.toString() + " for service type " + serviceToAddClass);
+                }
+            }
+        }
+
+        // Ensure that all Prioritizers are instantiate-able and that any load balancing configuration is correct
```
Comment (greyp9) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/groups/StandardProcessGroupSynchronizer.java:954]: consistency with "instantiable"

**Review by greyp9 — COMMENTED** [review]
Commented on a few minor typos.  Also did some testing:
- standalone NiFi looks fine (stops and starts tested)
- five node cluster (1:1 embedded zookeeper; small flow with ExecuteProcess)
- use Cluster UI to disconnect and reconnect cluster nodes
- forcibly shutdown multiple nodes, and restart; recovery successful, with normal exceptions
- noted instances of logs like following associated with cluster disconnect/reconnect (static flow, these also logged when running main branch):

```
Process Group with ID 44c3c9b4-8ef9-36d0-9e14-429e8e1fec21 exists in Proposed Flow but not in Currently Loaded Flow
Processor has a Scheduled State of RUNNING in Proposed Flow but ENABLED in Currently Loaded Flow
Processor has a Scheduled State of RUNNING in Proposed Flow but ENABLED in Currently Loaded Flow
```

Logged NIFI-9389 for small issue unrelated to this PR.

**Review by gresockj — CHANGES_REQUESTED** [review]
This will be a huge win for NiFi!  I love being able to make modifications to the flow while nodes are disconnected.  I tested out several things, including:
- Loading a 7000+ processor flow on a 3-node cluster
- Bringing down 1 or 2 nodes, both gracefully and non-gracefully, then deleting large parts of the flow, and finally bringing back up the nodes, which sync'ed perfectly
- Version controlling large flows, including root level controller services and parameter contexts.  Importing versioned flows from NiFi-registry.

Most of this worked very well, but I noticed some issues with synchronizing from registry with parameter context inheritance.
1. I noted below a place where `withParameterContextResolution` is needed to wrap a call.
2. I created a simple flow with parameter context inheritance.  Saving the flow to registry works and preserves the inheritance structure.  However, if I delete the parameter context attached to the versioned process group, reconstructing it fails.  I see this in the logs:
```
2021-11-18 07:13:41,639 INFO [NiFi Web Server-27] o.a.n.g.StandardProcessGroupSynchronizer Updating StandardProcessGroup[identifier=32f7836d-017d-1000-ffff-fffffd1dde0f,name=PC test] to VersionedFlowSnapshot[flowId=3a915b83-7d96-4aed-8c56-00638ff2a391, flowName=PC test, version=2, comments=Committing with PC update]; there are 2 differences to take into account:
Processor with ID 75e1b327-585c-34c2-896e-4567d4d6c481 exists in Proposed Flow but not in Currently Loaded Flow
Parameter Context Changed for Process Group with ID 52bc7024-3970-3dc0-8f25-ee339f66c611; flow 'Proposed Flow' has value Inheritance; flow 'Currently Loaded Flow' has value null
```
What appears to happen is that the proposed flow does not get assigned the "Inheritance" parameter context, and then does not add the single processor in the versioned group.  This causes the imported group to immediately be out of date with version control.  In the main branch, NiFi reconstructs the deleted "Inheritance" parameter context and successfully assigns it to the process group. I will post the flow I'm using on the JIRA ticket so this can be reproduced.

Once these kinks are worked out, everything else looks good to me.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/flow/PopularVoteFlowElection.java (hunk 1 above)
Comment (markap14) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/flow/PopularVoteFlowElection.java:145]: We can definitely remove the instance from this class. The class is still used elsewhere, though.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster-protocol/src/main/java/org/apache/nifi/cluster/protocol/StandardDataFlow.java [hunk 3]
```diff
@@ -136,4 +152,29 @@ public void error(final SAXParseException e) {
...
+            throw new FlowSerializationException("Could not parse flow as a VersionedDataflow", e);
+        }
+    }
+
+    public boolean isXml() {
+        if (flow == null || flow.length == 0) {
+            return true;
+        }
+
+        return flow[0] == '<';
```
[the remark this answers is not in the document]
Comment (markap14) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster-protocol/src/main/java/org/apache/nifi/cluster/protocol/StandardDataFlow.java:178]: I don't believe it's worth trying to determine validity, etc. The file should never be touched outside of NiFi.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/StandardNiFiServiceFacade.java [hunk 4]
```diff
@@ -2559,7 +2559,7 @@ public void verifyCanInstantiate(final String groupId, final FlowSnippetDTO snip
 
     @Override
     public void verifyComponentTypes(final VersionedProcessGroup versionedGroup) {
-        controllerFacade.verifyComponentTypes(versionedGroup);
+//        controllerFacade.verifyComponentTypes(versionedGroup);
```
[the remark this answers is not in the document]
Comment (markap14) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/StandardNiFiServiceFacade.java:2562]: Yeah intent was to remove the function all together. If there are types that are not available it should ghost the components. Just commented out to test behavior and forgot to remove it. Could potentially remove the method, but I can envision a future in which we want to perform some other sort of verification so left the method there as a placeholder.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/groups/StandardProcessGroupSynchronizer.java (hunk 2 above)
Comment (markap14) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/groups/StandardProcessGroupSynchronizer.java:954]: "instantiate-able"? I must have been tired :)

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/ProcessorNode.java [hunk 5]
```diff
@@ -174,7 +174,7 @@ public ScheduledState getScheduledState() {
      * states such as STOPPING and STARTING.
      *
      * @return the physical state of this processor [DISABLED, STOPPED, RUNNING,
-     *         STARTIING, STOPIING]
+     *         STARTING, STOPIING]
```
[the remark this answers is not in the document]
Comment (markap14) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/ProcessorNode.java:177]: Haha wow! I fixed the typo in STARTIING and completed missed the typo right next to it :)

**Comment by markap14** [comment]
Thanks for the reviews & testing all! I pushed a new commit (or two) and rebased against main to resolve the conflict. Newest version of the PR, I believe, addresses all concerns.

**Review by gresockj — APPROVED** [review]
Great, thanks for these updates, @markap14!  I tested the inherited Parameter Contexts with the latest commit, and everything works as expected now.

+1 from me.

**Review by greyp9 — APPROVED** [review]
Thanks for the updates.  LGTM

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for the updates to `FingerprintFactory` usage @markap14, that looks good.

Even with increasing system-tests timeouts to 90 minutes, it looks like they still failed to complete.  The main ci-workflow is 120 minutes, so you could try that. I would like to see at least one successful system-tests run on GitHub, otherwise we should disable the systems tests for GitHub and evaluate alternative solutions.

**Review by exceptionfactory — APPROVED** [review]
Thanks for tracking down the issues with the system-tests @markap14, green on all platforms! The addition of the troubleshooting logs for system-tests is also helpful.

Based on the current approvals and successful build, +1 merging!

**Pull request closed without merging**